finality-aleph/src/abft/current/performance/scorer.rs

```diff
@@ -0,0 +1,155 @@
+use current_aleph_bft::{NodeCount, NodeMap, Round};
+
+use crate::{abft::current::performance::Batch, UnverifiedHeader};
+
+pub type Score = Vec<Round>;
+
+/// Scoring ABFT performance based on returned ordered unit batches.
+pub struct Scorer {
+    newest_unit_by: NodeMap<Round>,
+}
+
+impl Scorer {
+    /// Create a new scorer for the provided node count.
+    pub fn new(node_count: NodeCount) -> Self {
+        Scorer {
+            newest_unit_by: NodeMap::with_size(node_count),
+        }
+    }
+
+    /// Add a batch of ordered units and return a score consisting of numbers of rounds a specific
+    /// node is behind.
+    pub fn process_batch<UH: UnverifiedHeader>(&mut self, batch: Batch<UH>) -> Score {
+        let max_round = batch.last().expect("batches always contain a head").round;
+        for unit in batch {
+            // Units are always added in order, so any unit created by an honest node
+            // here has a round greater than any that was included earlier.
+            // This is not necessarily true for forkers, but punishing them is fine.
+            self.newest_unit_by.insert(unit.creator, unit.round)
+        }
+        self.newest_unit_by
+            .size()
+            .into_iterator()
+            .map(|node_id| {
+                self.newest_unit_by
+                    .get(node_id)
+                    // All other units have lower round than head, so the saturating_sub is just
+                    // subtraction.
+                    .map(|unit_round| max_round.saturating_sub(*unit_round))
+                    // If we don't have a unit it's the same as having a unit of round -1.
```

Comment:
This comment seems off. If we do have a unit of round - 1 then the score won't be max round + 1 unless last unit was a genesis unit.

Reply:
This is supposed to refer to an absolute round -1 (which doesn't really exist, but conceptually), not `max_round-1`, although I don't know how to clarify this. Maybe something like "...as having a virtual unit of round -1." or "We can conceptualize not having a unit as having an impossible unit of round -1."?

Reply:
Nah...I get it now 🤦 to deconfuse the sentence it's enough to change the end to "of round equal to '-1'".